src/bin/ambit/cmd.rs

```diff
@@ -52,24 +53,122 @@ fn is_symlinked(link_name: &Path, target: &Path) -> bool {
+            }
+            // Strip prefix from all paths.
+            valid_paths = valid_paths
+                .iter()
+                .map(|path| -> AmbitResult<PathBuf> {
+                    Ok(path.strip_prefix(&start_path)?.to_path_buf())
+                })
+                .collect::<AmbitResult<Vec<PathBuf>>>()?;
+            paths.append(&mut valid_paths);
+        } else {
+            paths.push(PathBuf::from(&entry));
+        }
+    }
+    Ok(paths)
+}
+
 // Return iterator over path pairs in the form of `(repo_file, host_file)` from given entry.
-fn get_ambit_paths_from_entry<'a>(
-    entry: &'a Entry,
-) -> Box<dyn Iterator<Item = (AmbitPath, AmbitPath)> + 'a> {
-    Box::new(
-        entry
-            .left
-            .into_iter()
-            // If entry.right is None, entry.left is considered to be both repo and host path.
-            .zip(entry.right.as_ref().unwrap_or(&entry.left).into_iter())
-            .map(|(repo_path, host_path)| {
-                // Wrap the given paths as AmbitPath.
-                (
-                    AmbitPath::new(AMBIT_PATHS.repo.path.join(repo_path), AmbitPathKind::File),
-                    AmbitPath::new(AMBIT_PATHS.home.path.join(host_path), AmbitPathKind::File),
-                )
-            }),
-    )
+fn get_ambit_paths_from_entry(entry: &Entry) -> AmbitResult<Vec<(AmbitPath, AmbitPath)>> {
+    let (left_paths, right_paths): (Vec<PathBuf>, Option<Vec<PathBuf>>) =
+        if let Some(entry_right) = &entry.right {
+            (
+                get_paths_from_spec(&entry.left, PathBuf::from(AMBIT_PATHS.repo.to_str()?))?,
+                Some(get_paths_from_spec(
+                    &entry_right,
+                    PathBuf::from(AMBIT_PATHS.home.to_str()?),
+                )?),
+            )
+        } else {
+            // The right entry does not exist. Treat the left entry as both the repo and host paths.
+            (
+                get_paths_from_spec(&entry.left, PathBuf::from(AMBIT_PATHS.home.to_str()?))?,
+                None,
+            )
+        };
+    // The number of left and right paths may be different due to pattern matching.
+    // An error is thrown if they have different sizes.
+    if let Some(right_paths) = &right_paths {
+        if left_paths.len() != right_paths.len() {
+            // Format the vector of PathBuf as a string delimited by a newline.
+            let format_paths = |paths: &Vec<PathBuf>| {
+                paths
+                    .iter()
+                    .map(|path| path.as_path().display().to_string())
+                    .collect::<Vec<String>>()
+                    .join("\n")
+            };
+            return Err(AmbitError::Other(format!(
+                "Entry has imbalanced left and right side due to pattern matching\nAttempted to sync:\n{}\nWITH:\n{}",
+                format_paths(&left_paths), format_paths(right_paths),
+            )));
+        }
+    }
+    Ok(left_paths
```

Comment:
It may be more useful to simply use a `for` loop here to avoid cloning `left_paths`.